File: 01_projects/code/java-playwright/cbaps-dawms/src/test/java/com/cbaps/pages/RequisitionPage.java
package com.cbaps.pages;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import com.cbaps.library.PlaywrightManager;

/**
 * RequisitionPage: Create and manage requisitions in CBAPS
 * Handles requisition title, fund type, and navigation to funding lines
 */
public class RequisitionPage {

	private static final Logger log = LoggerFactory.getLogger(RequisitionPage.class);
	private final Page page;
	private final PlaywrightManager pwm;

	// Locators
	private final Locator titleInput;
	private final Locator fundTypeDropdown;
	private final Locator submitButton;
	private final Locator statusBadge;
	private final Locator goToFundingLink;
	private final Locator routeForApprovalButton;

	public RequisitionPage(Page page, PlaywrightManager pwm) {
		this.page = page;
		this.pwm = pwm;

		// Initialize locators
		this.titleInput = page.locator("#requisitionTitle");
		this.fundTypeDropdown = page.locator("#fundType");
		this.submitButton = page.locator("button:has-text('Submit')");
		this.statusBadge = page.locator("#reqStatus");
		this.goToFundingLink = page.locator("a:has-text('Funding Lines')");
		this.routeForApprovalButton = page.locator("button:has-text('Route for Approval')");

		// Stability anchor: wait for page to be ready
		pwm.waitVisible("#requisitionTitle");
		log.info("Requisition Page loaded successfully.");
	}

	/**
	 * Create a new requisition with title and fund type
	 * @param title - Requisition title
	 * @param fundType - Fund type selection
	 */
	public void createRequisition(String title, String fundType) {
		pwm.type(titleInput, title);
		log.info("Entered requisition title: " + title);
		
		fundTypeDropdown.selectOption(fundType);
		log.info("Selected fund type: " + fundType);
		
		pwm.click(submitButton);
		log.info("Clicked Submit button.");
	}

	/**
	 * Get current requisition status
	 * @return Status text
	 */
	public String getStatus() {
		String status = statusBadge.textContent();
		log.info("Retrieved status: " + status);
		return status;
	}

	/**
	 * Navigate to Funding Lines page
	 * @return FundingLinesPage
	 */
	public FundingLinesPage goToFundingLines() {
		pwm.click(goToFundingLink);
		log.info("Navigated to Funding Lines page.");
		return new FundingLinesPage(page, pwm);
	}

	/**
	 * Route requisition for approval
	 * @return RoutingApprovalPage
	 */
	public RoutingApprovalPage routeForApproval() {
		pwm.click(routeForApprovalButton);
		log.info("Clicked 'Route for Approval' button.");
		return new RoutingApprovalPage(page, pwm);
	}
}

File: 01_projects/code/python-playwright/cbaps-dawms/basic-pom/cbaps/pages/requisition_page.py
"""
RequisitionPage: Create and manage requisitions in CBAPS
Handles requisition title, fund type, and navigation to funding lines
"""

import logging
from playwright.sync_api import Page, Locator
from cbaps.library.playwright_manager import PlaywrightManager

logger = logging.getLogger(__name__)


class RequisitionPage:
    """
    RequisitionPage handles requisition creation workflow.
    This is the Python equivalent of the Java RequisitionPage.
    """
    
    def __init__(self, page: Page, pwm: PlaywrightManager):
        """
        Initialize RequisitionPage
        
        Args:
            page: Playwright Page object
            pwm: PlaywrightManager instance
        """
        self.page = page
        self.pwm = pwm
        
        # Locators
        self.title_input: Locator = page.locator("#requisitionTitle")
        self.fund_type_dropdown: Locator = page.locator("#fundType")
        self.submit_button: Locator = page.locator("button:has-text('Submit')")
        self.status_badge: Locator = page.locator("#reqStatus")
        self.go_to_funding_link: Locator = page.locator("a:has-text('Funding Lines')")
        self.route_for_approval_button: Locator = page.locator(
            "button:has-text('Route for Approval')"
        )
        
        # Stability anchor: wait for page to be ready
        pwm.wait_visible("#requisitionTitle")
        logger.info("Requisition Page loaded successfully.")
    
    def create_requisition(self, title: str, fund_type: str):
        """
        Create a new requisition with title and fund type
        
        Args:
            title: Requisition title
            fund_type: Fund type selection
        """
        self.pwm.type(self.title_input, title)
        logger.info(f"Entered requisition title: {title}")
        
        self.fund_type_dropdown.select_option(fund_type)
        logger.info(f"Selected fund type: {fund_type}")
        
        self.pwm.click(self.submit_button)
        logger.info("Clicked Submit button.")
    
    def get_status(self) -> str:
        """
        Get current requisition status
        
        Returns:
            Status text
        """
        status = self.status_badge.text_content()
        logger.info(f"Retrieved status: {status}")
        return status
    
    def go_to_funding_lines(self) -> 'FundingLinesPage':
        """
        Navigate to Funding Lines page
        
        Returns:
            FundingLinesPage instance
        """
        from cbaps.pages.funding_lines_page import FundingLinesPage
        
        self.pwm.click(self.go_to_funding_link)
        logger.info("Navigated to Funding Lines page.")
        return FundingLinesPage(self.page, self.pwm)
    
    def route_for_approval(self) -> 'RoutingApprovalPage':
        """
        Route requisition for approval
        
        Returns:
            RoutingApprovalPage instance
        """
        from cbaps.pages.routing_approval_page import RoutingApprovalPage
        
        self.pwm.click(self.route_for_approval_button)
        logger.info("Clicked 'Route for Approval' button.")
        return RoutingApprovalPage(self.page, self.pwm)
